### PR TITLE
JavaScript: Add new query `InvalidEntityTranscoding`.

### PR DESCRIPTION
**Pull request by xiemaisi**
A simple, lightweight query that spots a common mistake people make when writing HTML entity encoders/decoders: when encoding, `&` has to be encoded first to avoid double-encoding ampersands introduced by the encoding of other characters; conversely, when decoding it has to be decoded last to avoid the decoded ampersand being interpreted as part of an entity reference later on.

Finds good [results](https://lgtm.com/query/1969458219962749443/) on LGTM.com, [here](https://git.semmle.com/gist/max/b189e2e6d42ba947c5d09495dea41129) is the full report (internal link). [One](https://github.com/mout/mout/blob/master/src/string/unescapeHtml.js#L7) of the results is in a moderately popular (but not very actively maintained) [utility-belt library](https://www.npmjs.com/package/mout).

@esben-semmle suggested looking for a similar problem with URL encoding and `%`, but a quick exploratory query seems to indicate that people don't often implement URL transcoding by hand.

### CODE REVIEW

**Comment by xiemaisi** [comment]
No particular rush to review this, it's not going into 1.19.

File: javascript/ql/src/Security/CWE-116/InvalidEntityTranscoding.ql [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+DataFlow::Node getASimplePredecessor(DataFlow::Node nd) {
+  result = nd.getAPredecessor() and
+  not nd.(DataFlow::SsaDefinitionNode).getSsaVariable().getDefinition() instanceof SsaPhiNode
+}
+
+/**
+ * A call to `String.prototype.replace` that replaces all instances of a pattern.
+ */
+class Replacement extends DataFlow::Node {
+  RegExpLiteral pattern;
```
Comment (asger-semmle) [inline javascript/ql/src/Security/CWE-116/InvalidEntityTranscoding.ql]: I take it string patterns here will always be reported by IncompleteSanitization, and that's why we don't include them here? Maybe worth leaving a comment about this.

Comment (xiemaisi) [reply javascript/ql/src/Security/CWE-116/InvalidEntityTranscoding.ql]: Correct; I'll add a note.

**Comment by xiemaisi** [comment]
Ping @mc-semmle for doc review (but https://github.com/Semmle/ql/pull/555 is more urgent).

**Comment by esbena** [comment]
LGTM too, but I see two features that could go in a later PR.

1. 
How about escapes with backslashes and the escaping of the backslash it self?

2.
`getPreviousReplacement` seems relevant for:

https://github.com/Semmle/ql/blob/d31c9950f9a9dec68d06c2d7af78761520879871/javascript/ql/src/Security/CWE-116/IncompleteSanitization.ql#L88-L94

**Comment by xiemaisi** [comment]
> How about escapes with backslashes and the escaping of the backslash it self?

That's a very interesting idea; let me play with that a little bit.

File: javascript/ql/src/Security/CWE-116/InvalidEntityTranscoding.qhelp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+<p>
+The decoding function, however, incorrectly decodes <code>&amp;amp;</code> into <code>&amp;</code>
+before handling the other characters. So while it correctly decodes the first example above,
+it decodes the second example (<code>&amp;quot;</code>) to <code>&quot;</code> (a single double quote),
+which is not correct.
+</p>
+
+<p>
+Instead, the decoding function should decode ampersand last:
```
Comment (mc-semmle) [inline javascript/ql/src/Security/CWE-116/InvalidEntityTranscoding.qhelp]: Tiny suggestion:
Instead, the decoding function should decode THE ampersand last:

**Comment by xiemaisi** [comment]
I have significantly rewritten the query based on @esben-semmle's suggestion, generalising it beyond HTML transcoding to a few other kinds of (un-)escaping.

I like [the new results](https://lgtm.com/query/7499871942373142801/) a lot, for instance [this one](https://github.com/jashkenas/underscore/blob/eeea70c457db3aebaef5b9251661f779844758fe/underscore.js#L903) from an old version of underscore. Full comparison is running.

File: javascript/ql/src/Security/CWE-116/DoubleEscaping.ql [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      if pred.escapes(_, metachar) then
+        result = pred
+      else
+        result = pred.getAnEarlierEscaping(metachar)
+    )
+  }
+
+  /**
+   * Gets an earlier replacement in this chain of replacements that
+   * performs a unescaping.
```
Comment (esben-semmle) [inline javascript/ql/src/Security/CWE-116/DoubleEscaping.ql]:
```suggestion
   * performs an unescaping.
```

**Comment by xiemaisi** [comment]
Results from full run look good as well: [internal link](https://git.semmle.com/gist/max/273a9be9d2cffa631610b2ad84a167d3).

**Comment by xiemaisi** [comment]
Rebased to resolve conflict on change notes.

Perhaps @mc-semmle wants to take another look at the query help, which has been rewritten to reflect the expanded scope of the query (but no rush; it's not relevant for 1.19).

**Comment by mc-semmle** [comment]
Documentation LGTM, I've approved the changes :)

**Comment by xiemaisi** [comment]
Thanks, @mc-semmle! I've resolved the conflict on the change notes, so this should be good to go in once it's green again.

**Pull request merged**